### PR TITLE
decision: frontend documentation

### LINKED ISSUE #128: Write decision about Frontend Documentation system
As suggested by @Bushuo and @badnames (meeting notes and https://github.com/ElektraInitiative/PermaplanT/pull/104#issuecomment-1475354889)

- [ ] I did not yet look deeper into it, but maybe [React Docgen](https://github.com/reactjs/react-docgen/blob/5.x/README.md#flow-and-typescript-support) could be combined with mdBooks.
- [ ] React Docgen outputs a json file that we could then parse into a markdown file.
- [ ] https://docusaurus.io/
- [ ]   [API-Extractor](https://api-extractor.com/)
- [ ]   [TypeDoc](https://typedoc.org)
- [ ] JSDoc
    - https://jsdoc.app/about-getting-started.html
    - https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#param-and-returns
    - https://www.inkoop.io/blog/a-guide-to-js-docs-for-react-js/
    - https://github.com/jsdoc/jsdoc
- [ ] Storybook
- [ ] React Styleguidist

### PR DESCRIPTION
**Pull request by Bushuo**
Closes #128

## Basics

- [X] Details of what you changed are in commit messages
      (first line should have `module: short statement` syntax)
- [X] References to issues, e.g. `close #X`, are in the commit messages.
- [X] The buildserver is happy.
- [X] The PR is rebased with current master.

## Checklist

- [ ] I added unit tests for my code
- [ ] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation
- [ ] I fixed all affected decisions
- [ ] I added code comments, logging, and assertions as appropriate
- [ ] I mentioned [every code](/.reuse/dep5) not directly written by me in [reuse syntax](https://reuse.software/)

## Review

- [ ] Documentation is conforming to [our Documentation Guidelines](/doc/documentation.md)
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to our Coding Guidelines
- [ ] Code is consistent to [our Design Decisions](/doc/decisions)

### CODE REVIEW

**Comment by Bushuo** [comment]
I could include screenshots of how the embedded documentation looks in my local setup, if you want to see :)

**Review by absurd-turtle — APPROVED** [review]

**Review by ThorbenStaufer — APPROVED** [review]

**Review by buenaflor — APPROVED** [review]
👍

File: doc/decisions/frontend_documentation.md [hunk 1]
```diff
@@ -0,0 +1,96 @@
+# Frontend Documentation
+
+## Problem
+
+We need to choose a workflow, tooling that helps us to keep the documentation up to date.
+An easy to read and up to date documentation reduces the entry barrier for new-comers.
```
Comment (markus2330) [inline doc/decisions/frontend_documentation.md:6]: If it refers to a noun, "-" are needed.

```suggestion
An easy to read and up-to-date documentation reduces the entry barrier for newcomers.
```

File: doc/decisions/frontend_documentation.md [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+Without proper documentation maintenance becomes increasingly more difficult as the project grows.
+
+## Constraints
+
+1. The documentation should be easy to navigate.
+2. The documentation should for the most part, be automatically extracted from the source code.
```
Comment (markus2330) [inline doc/decisions/frontend_documentation.md:12]:
```suggestion
2. The documentation should for the most part, be automatically extracted from the source code.
3. The documentation should be searchable.
```

File: doc/decisions/frontend_documentation.md [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+## Assumptions
+
+1. With a good and beautiful documentation tool, developers are more likely to use and maintain it.
+2. With a documentation tool that enables the viewing of react components, the code quality is improved. Presentation components, should not contain logic with side effects.
```
Comment (markus2330) [inline doc/decisions/frontend_documentation.md:17]:
```suggestion
2. With a documentation tool that enables the viewing of react components, the code quality is improved. 
Presentation components, should not contain logic with side effects.
```

File: doc/decisions/frontend_documentation.md [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+Hooks should be used outside of presentation components, with props passed into them.
```
Comment (markus2330) [inline doc/decisions/frontend_documentation.md:18]: This should be in general guidelines about frontend (architecture), and we should like from here to this document.

File: doc/decisions/frontend_documentation.md [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+If those rules are not followed, it will be hard to display the components inside the documentation.
+3. With an overview of all presentation components, the whole team can instantly see what is available.
+
+## Considered Alternatives (documentation extraction)
+
+- API-Extractor: [API-Extractor](https://api-extractor.com/) is a tool to extract the public API of a typescript project. It is more suited for libraries.
+
+
```
Comment (markus2330) [inline doc/decisions/frontend_documentation.md:26]:
```suggestion

```

File: doc/decisions/frontend_documentation.md [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+- React Docgen: [React Docgen](https://github.com/reactjs/react-docgen) is a tool from the React Community to extract information from React Components for documentation generation purposes.
+It is not actively maintained anymore.
+It does not generate markdown files, so an additional step is needed, to get just the most basic information.
```
Comment (markus2330) [inline doc/decisions/frontend_documentation.md:29]:
```suggestion
  - It is not actively maintained anymore.
  - It does not generate markdown files, so an additional step is needed, to get just the most basic information.
```

File: doc/decisions/frontend_documentation.md [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+- JSDoc: [JSDoc](https://jsdoc.app/) is an API documentation generator for JavaScript.
+It is more suited for JavaScript.
+
+## Decision
+We will use [TypeDoc](https://typedoc.org/) with the markdown plugin to extract documentation from the source code.
+
+## Rationale
+TypeDoc has automatic support for most Tags of JSDoc via the TypeScript compiler.
+Additional Tags are supported as well.
+It is easy to configure and invoke and extensible via plugins.
```
Comment (markus2330) [inline doc/decisions/frontend_documentation.md:40]: Say more about which plugins we will use/need. Is something missing?

```suggestion

- TypeDoc has automatic support for most Tags of JSDoc via the TypeScript compiler.
- Additional Tags are supported as well.
- It is easy to configure and invoke and extensible via plugins.
```

File: doc/decisions/frontend_documentation.md [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+
+## Considered Alternatives (viewing components)
+
+- React Styleguidist: [React Styleguidist](https://react-styleguidist.js.org/)
+Is a popular and maintained solution to develop a project wide style-guide in React.
```
Comment (markus2330) [inline doc/decisions/frontend_documentation.md:45]:
```suggestion
Is a popular and maintained solution to develop a project-wide style guide in React.
```

File: doc/decisions/frontend_documentation.md [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+## Decision
+We will use [TypeDoc](https://typedoc.org/) with the markdown plugin to extract documentation from the source code.
+
+## Rationale
+TypeDoc has automatic support for most Tags of JSDoc via the TypeScript compiler.
+Additional Tags are supported as well.
+It is easy to configure and invoke and extensible via plugins.
+
+## Considered Alternatives (viewing components)
+
```
Comment (markus2330) [inline doc/decisions/frontend_documentation.md:43]: Docusaurus etc. should be also mentioned. Basically everything that we know and could somewhat do the job should be mentioned.

File: doc/decisions/frontend_documentation.md [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+- React Styleguidist: [React Styleguidist](https://react-styleguidist.js.org/)
+Is a popular and maintained solution to develop a project wide style-guide in React.
+It has fewer features than Storybook.
+It is hard to setup, as it assumes the usage of Create React App.
+
+## Decision
+We will use [Storybook](https://storybook.js.org/) to enable a project wide component overview and style-guide.
+It has support for `MDX`, a format that enables embeddable components in markdown.
+It has support for automated testing via `Jest` and `Playwright`.
```
Comment (markus2330) [inline doc/decisions/frontend_documentation.md:52]: Maybe we can also immediately decide on that?

**Review by markus2330 — CHANGES_REQUESTED** [review]
Very good first version. I personally would split this into two decisions and link them to each other (or even three to decide about testing tool).

File: doc/decisions/frontend_documentation.md (hunk 10 above)
Comment (markus2330) [reply doc/decisions/frontend_documentation.md:52]: (But maybe in a separate PR because it seems like we already have consensus about the rest.)

**Comment by markus2330** [comment]
Yes, if you already created screenshots it would be nice to see.

**Review by badnames — APPROVED** [review]

**Comment by Bushuo** [comment]
Hi @markus2330 
> Yes, if you already created screenshots it would be nice to see.

These are two screenshots from my local setup. I changed how the [typedoc-plugin-markdown](https://www.npmjs.com/package/typedoc-plugin-markdown) renders the markdown files, such that they can be recognized by storybook directly. There is still some work to do, to make the linking and hierarchical structure of the generated files work. I guess it is a good start.
README_md
Enumeration_Quality_md

**Comment by Bushuo** [comment]
The local storybook setup, I have been working on, now displays all generated typedoc documentation files, links between them, and search 🎉🎉
Screenshot 2023-03-28 at 23 51 36

**Comment by markus2330** [comment]
Great to see it! Please fix the remaining small things so that we can get this merged!

File: doc/decisions/frontend_documentation.md (hunk 4 above)
Comment (Bushuo) [reply doc/decisions/frontend_documentation.md]: @markus2330 I added it to `contrib/frontend`.

File: doc/decisions/frontend_documentation.md [hunk 11]
```diff
@@ -0,0 +1,99 @@
+# Frontend Documentation
+
+## Problem
+
+We need to choose a workflow, tooling that helps us to keep the documentation up to date.
+An easy to read and up-to-date documentation reduces the entry barrier for newcomers.
```
Comment (markus2330) [inline doc/decisions/frontend_documentation.md:6]:
```suggestion
An easy-to-read and up-to-date documentation reduces the entry barrier for newcomers.
```

File: doc/decisions/frontend_documentation.md [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+Without proper documentation maintenance becomes increasingly more difficult as the project grows.
+
+## Constraints
+
+1. The documentation should be easy to navigate.
+2. The documentation should for the most part, be automatically extracted from the source code.
```
Comment (markus2330) [inline doc/decisions/frontend_documentation.md:12]:
```suggestion
2. The documentation should, for the most part, be automatically extracted from the source code.
```

File: doc/decisions/frontend_documentation.md [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+3. The documentation should be searchable.
+
+## Assumptions
+
+1. With a good and beautiful documentation tool, developers are more likely to use and maintain it.
+2. With a documentation tool that enables the viewing of react components, the code quality is improved.
+If the rules about component design in [contrib/frontend](../contrib/frontend.md#component-design) are not followed, it will be hard to display the components inside the documentation.
+3. With an overview of all presentation components, the whole team can instantly see what is available.
```
Comment (markus2330) [inline doc/decisions/frontend_documentation.md:20]:
```suggestion
3. The rules about component design in [contrib/frontend](../contrib/frontend.md#component-design) are followed.
4. With an overview of all presentation components, the whole team can instantly see what is available.
```

File: doc/decisions/frontend_documentation.md [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+
+- JSDoc: [JSDoc](https://jsdoc.app/) is an API documentation generator for JavaScript.
+It is more suited for JavaScript.
+
+## Decision
+We will use [TypeDoc](https://typedoc.org/) with the markdown plugin to extract documentation from the source code.
+
+## Rationale
+
+- TypeDoc has automatic support for most Tags of JSDoc via the TypeScript compiler.
```
Comment (markus2330) [inline doc/decisions/frontend_documentation.md:37]:
```suggestion
- TypeDoc has automatic support for most tags of JSDoc via the TypeScript compiler.
```

File: doc/decisions/frontend_documentation.md [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+- Additional Tags are supported as well.
```
Comment (markus2330) [inline doc/decisions/frontend_documentation.md:38]:
```suggestion
- Additional tags are supported as well.
```

File: doc/decisions/frontend_documentation.md [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+- It is easy to configure and invoke and extensible via plugins.
```
Comment (markus2330) [inline doc/decisions/frontend_documentation.md:39]:
```suggestion
- It is easy to configure and invoke.
- It is extensible via plugins.
```

File: doc/decisions/frontend_documentation.md (hunk 10 above)
Comment (markus2330) [reply doc/decisions/frontend_documentation.md:55]: This is still open. Can you please comment on #39.

File: doc/decisions/frontend_documentation.md [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+      canvas: { hidden: true },
+    },
+  }}
+/>
+
+<Md />
+```
+
+### How to structure a storybook
+https://storybook.js.org/blog/structuring-your-storybook/
```
Comment (markus2330) [inline doc/decisions/frontend_documentation.md:98]: This should then also be in the main documentation.

**Review by markus2330 — APPROVED** [review]
Great job, a few minor last suggestions. Please (automatically) reformat this Markdown file so that it is consistent, see #159. Please tell everyone about this decision next meeting.

File: doc/decisions/frontend_documentation.md (hunk 17 above)
Comment (Bushuo) [reply doc/decisions/frontend_documentation.md:99]: I added it to the Helpful Links section

**Comment by markus2330** [comment]
**Great** job! Looking forward to see this realized.

**Pull request merged**